### PR TITLE
Add change tracks modal

### PR DESCRIPTION
**Pull request by kntsoriano**

### CODE REVIEW

File: app/controllers/temp/mentored_tracks_controller.rb [hunk 1]
```diff
@@ -0,0 +1,19 @@
+module Temp
```
Comment (kntsoriano) [inline app/controllers/temp/mentored_tracks_controller.rb]: @iHiD I've added this temporary controller to get and update the user's mentored tracks.

File: app/commands/solution/mentor_request/retrieve_tracks.rb [hunk 2]
```diff
@@ -3,9 +3,8 @@ class MentorRequest
     class RetrieveTracks
```
Comment (kntsoriano) [inline app/commands/solution/mentor_request/retrieve_tracks.rb]: I removed `selected_id` here. We handle passing the selected track for the mentoring queue differently now.

File: app/helpers/react_components/mentoring/queue.rb [hunk 3]
```diff
@@ -13,10 +13,14 @@ def to_s
         super(
           "mentoring-queue",
           {
-            request: request,
-            tracks: track_data,
-            exercises: exercise_data,
-            sort_options: SORT_OPTIONS
+            queue_request: queue_request,
+            tracks_request: tracks_request,
+            default_track: default_track,
```
Comment (kntsoriano) [inline app/helpers/react_components/mentoring/queue.rb:18]: This is how you choose the selected track.

File: app/helpers/react_components/mentoring/queue.rb [hunk 4]
```diff
@@ -31,25 +35,34 @@ def to_s
...
+            }
           }
         }
       end
 
-      memoize
-      def track_data
-        ::Solution::MentorRequest::RetrieveTracks.(mentor)
+      def default_track
+        track = track_data.first
```
Comment (iHiD) [inline app/helpers/react_components/mentoring/queue.rb:52]:
```suggestion
        # TODO: This should be whatever the mentor last viewed. 
        # Store that value somewhere and retrieve it here.
        track = track_data.first
```

**Review by iHiD — CHANGES_REQUESTED** [review]
Reviewed via loom: https://www.loom.com/share/7905eba8b0824e15bf142e0cf70866b5

File: app/controllers/temp/mentored_tracks_controller.rb (hunk 1 above)
Comment (iHiD) [reply app/controllers/temp/mentored_tracks_controller.rb]: Cool. I've added `skip_before_action :verify_authenticity_token` to this as it wasn't working for me without.

File: app/helpers/react_components/mentoring/queue.rb [hunk 5]
```diff
@@ -31,25 +35,36 @@ def to_s
...
+            }
           }
         }
       end
 
-      memoize
-      def track_data
-        ::Solution::MentorRequest::RetrieveTracks.(mentor)
+      def default_track
+        # TODO: This should be whatever the mentor last viewed.
```
Comment (iHiD) [inline app/helpers/react_components/mentoring/queue.rb]:
```suggestion
        # TODO: This should be whatever the mentor last viewed.
```

File: app/commands/solution/mentor_request/retrieve_tracks.rb [hunk 6]
```diff
@@ -3,9 +3,8 @@ class MentorRequest
...
       include Mandate
 
-      def initialize(mentor, selected_id: nil)
+      def initialize(mentor)
```
Comment (ErikSchierboom) [inline app/commands/solution/mentor_request/retrieve_tracks.rb]: Could this now be replaced with `initialize_with :mentor`?

File: test/javascript/components/mentoring/Queue.test.js [hunk 7]
```diff
@@ -0,0 +1,333 @@
+import React from 'react'
+import { render, screen } from '@testing-library/react'
+import { rest } from 'msw'
+import { setupServer } from 'msw/node'
+import '@testing-library/jest-dom/extend-expect'
+import { Queue } from '../../../../app/javascript/components/mentoring/Queue'
+import { TestQueryCache } from '../../support/TestQueryCache'
+import { silenceConsole } from '../../support/silence-console'
+
```
Comment (ErikSchierboom) [inline test/javascript/components/mentoring/Queue.test.js:9]: It looks like there are only error-case tests. Did you intend to not add tests for the success case?

File: app/javascript/components/modals/MentorChangeTracksModal.tsx [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+export const MentorChangeTracksModal = ({
+  onClose,
+  tracks,
+  cacheKey,
+  links,
+  onSuccess,
+  ...props
+}: Omit<ModalProps, 'className'> & {
```
Comment (ErikSchierboom) [inline app/javascript/components/modals/MentorChangeTracksModal.tsx:24]: Using `Omit`, fancy!

File: app/javascript/components/modals/MentorChangeTracksModal.tsx [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+  links: Links
+  tracks: readonly MentoredTrack[]
```
Comment (ErikSchierboom) [inline app/javascript/components/modals/MentorChangeTracksModal.tsx:26]: Do you think that maybe we should be using `readonly` a lot more, as most of our props are read-only after all? I think we could use it for all properties here. The `Readonly<T>` type is something that we could consider using: https://www.typescriptlang.org/docs/handbook/utility-types.html#readonlytype

**Review by ErikSchierboom — APPROVED** [review]

**Comment by iHiD** [comment]
@kntsoriano I've standardised the `MentoredTrack` into a proper type with a serializer and used it in all the places that `MentoredTrack` and `Track` were being used in the mentoring bit. I've also rebased. I'll merge this when it's all green. Again, feel free to update anything you don't think I've got right :)

**Comment by iHiD** [comment]
@kntsoriano Also note there are a few comments from @ErikSchierboom here that should be responded to :)

File: app/commands/solution/mentor_request/retrieve_tracks.rb (hunk 6 above)
Comment (kntsoriano) [reply app/commands/solution/mentor_request/retrieve_tracks.rb]: Yes. I'll do this now.

File: app/javascript/components/modals/MentorChangeTracksModal.tsx (hunk 9 above)
Comment (kntsoriano) [reply app/javascript/components/modals/MentorChangeTracksModal.tsx:26]: I agree. I think we should do readonly almost everywhere. I'll apply this to my next PRs and see where we trip up.

**Pull request merged**